### PR TITLE
Cilium: Remove encryption state on NodeDelete events

### PR DESCRIPTION
**Pull request by jrfastab**
Delete encryption state when nodes are deleted otherwise the datapath will contain state for every node that existed.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7776)

### CODE REVIEW

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

File: pkg/datapath/linux/node.go [hunk 1]
```diff
@@ -732,6 +789,13 @@ func (n *linuxNodeHandler) NodeConfigurationChanged(newConfig datapath.LocalNode
 
 	n.updateOrRemoveNodeRoutes(prevConfig.AuxiliaryPrefixes, newConfig.AuxiliaryPrefixes)
 
+	if !newConfig.EnableIPSec {
+		err := n.removeEncryptRules()
```
Comment (joestringer) [inline pkg/datapath/linux/node.go:807]: Does this succeed if encryption was previously not enabled?

**Review by joestringer — APPROVED** [review]

**Comment by joestringer** [comment]
Failed key allocation attempts in the test `Check connectivity with transparent encryption and VXLAN encapsulation`: 

```
07:23:56  K8sDatapathConfig Encapsulation 
07:23:56    Check connectivity with transparent encryption and VXLAN encapsulation
07:23:56    /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:410
07:25:45  STEP: Performing Cilium preflight check
07:25:45  STEP: Waiting for all etcd-operator pods are ready
07:29:39  === Test Finished at 2019-04-18T07:29:36Z====
07:29:39  ===================== TEST FAILED =====================
```
...
```
07:30:56  Number of "context deadline exceeded" in logs: 0
07:30:56  Number of "level=error" in logs: 0
07:30:56  ⚠️  Number of "level=warning" in logs: 34
07:30:56  Number of "Cilium API handler panicked" in logs: 0
07:30:56  ⚠️  Number of "Goroutine took lock for more than" in logs: 26
07:30:56  Top 5 errors/warnings:
07:30:56  Skipping symbol substitution
07:30:56  Key allocation attempt failed
07:30:56  Ongoing key allocation has been cancelled
07:30:56  Error changing endpoint identity
07:30:56  Creation of endpoint failed
```
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/11711/execution/node/84/log/

**Comment by jrfastab** [comment]
The failure appears to be the same issue failing on nightly. I'm going to mark wip and retry, maybe we can reliably reproduce here now.

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
OK looks like we can reproduce today.

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
Moving key secret out of per test logic up into BeforeAll section.

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
passed 1/1 with single secret updates

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
debug note: We are seeing XfrmOutStateMode errors in the logs, the primary reason for this error is the `skb->len >  dst_route_mtu` which pushes a EMSGSIZE error up the stack (from ip -> tcp layers) and should also send a ICMP "local" (unclear if this just adjust state or really sends a icmp with complete skb). My working hypothesis here is for some reason testsclient_ds, that seems to be failing here and in nightly due to context timeout, need to get an identity and the allocation logic is building a skb too large to fit in the route mtu but is not responding to the emsgsize error and icmp local event to craft a smaller skb and resend something that "will" work. In all my test beds testclient_ds and testds both come online with a proper identity and also don't have the XfrmOutStateMode counter incrementing.

Far fetched? Not sure need to device some way to test this but the state error seems to be correlated.

Going to try net-next as well now perhaps it will work and give another datapoint.

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
Another issue is I'm not always getting Delete events in GKE with multiple nodes resizing. This leaves state floating around but is unrelated to this patch.

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
@joestringer Added some more tests and a couple fixes that were blocking those tests from passing.

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 2]
```diff
@@ -310,6 +342,10 @@ func UpsertIPsecEndpoint(local, remote *net.IPNet, dir IPSecDir) (uint8, error)
 
 // DeleteIPsecEndpoint deletes a endpoint associated with the remote IP address
 func DeleteIPsecEndpoint(remote *net.IPNet) {
+	scopedLog := log.WithFields(logrus.Fields{
+		"remote-ip": remote.IP,
+	})
+	scopedLog.Info("Delete IPsec Endpoint")
```
Comment (joestringer) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]: How often will this occur? We're trying to get to a point where there is little to no `Info()` messages after startup. Maybe it's okay if this is likely to just be once per node add/remove event. (/cc @tgraf )

**Review by joestringer — APPROVED** [review]

File: pkg/datapath/linux/ipsec/ipsec_linux.go (hunk 2 above)
Comment (jrfastab) [reply pkg/datapath/linux/ipsec/ipsec_linux.go]: Good catch lets just remove it. Its not so useful except when trying to find delete events but its better to debug the nodeDelete and k8s watcher at that point.

**Comment by jrfastab** [comment]
net-next failed, https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/11778/

**Comment by jrfastab** [comment]
test-me-please

**Comment by aanm** [comment]
@jrfastab please mark this PR once it's ready for review

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test run failed because runtime tests tried to create an interface with mtu 65336 which results in RTNETLINK error. Added a patch to clamp the MTU as a workaround.

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

File: pkg/mtu/detect_linux.go [hunk 3]
```diff
@@ -65,6 +66,9 @@ func autoDetect() (int, error) {
 	}
 
 	if mtu := link.Attrs().MTU; mtu != 0 {
+		if mtu < maxCiliumMTU {
```
Comment (joestringer) [inline pkg/mtu/detect_linux.go]: isn't this the reverse of what the commit message is trying to say?

**Review by joestringer — CHANGES_REQUESTED** [review]

File: pkg/mtu/detect_linux.go (hunk 3 above)
Comment (jrfastab) [reply pkg/mtu/detect_linux.go]: typo thanks.

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
popping off the Delete logic on k8s events for now it seems to be causing failures.

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
failed, destroy/creating VMs 

16:38:58  ==> k8s2-1.10: VM not created. Moving on...
16:38:58  ==> k8s1-1.10: VM not created. Moving on...
16:38:58  ==> runtime: VM not created. Moving on.

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
the issue that was causing my failures is that we don't include the 'table' in the delete route logic so we can delete arbitrary routes that are needed.

**Comment by jrfastab** [comment]
Merged in PR https://github.com/cilium/cilium/pull/8048 after some clean up

**Pull request closed without merging**